packages/intrepid2/src/Projection/Intrepid2_ProjectionTools.hpp

```diff
@@ -816,20 +828,27 @@ class ProjectionTools {
 
         for(ordinal_type i=0; i<n; ++i) {
           for (ordinal_type ic = 0; ic < numCells; ic++)
-            serialBasisCoeffs(ic,elemDof(i)) = serialElemRhs(i,ic);
+            serialBasisCoeffs(ic,serialElemDof(i)) = serialElemRhs(i,ic);
         }
       }
       else {
         Kokkos::View<valueType**,Kokkos::LayoutLeft,HostDeviceType> pivVec("pivVec", 2*(m+n), 1);
         Kokkos::View<valueType**,Kokkos::LayoutLeft,HostDeviceType> serialElemRhs("serialElemRhs", n+m, 1 );
+        Kokkos::DynRankView<typename ViewType2::value_type, HostDeviceType> b_host("b_host",elemRhs.extent(1));
+        auto b_device = Kokkos::create_mirror_view(typename DeviceType::memory_space(), b_host);
+        Kokkos::DynRankView<typename ViewType2::value_type, HostDeviceType> basisCoeffsSubviewCopy_host("basisCoeffsSubviewCopy_host",basisCoeffs.extent(1));
+        auto basisCoeffsSubviewCopy_device = Kokkos::create_mirror_view(typename DeviceType::memory_space(), basisCoeffsSubviewCopy_host);
+
         for (ordinal_type ic = 0; ic < numCells; ic++) {
-          auto A = Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(),
-              Kokkos::subview(elemMat, ic, Kokkos::ALL(), Kokkos::ALL()));
-          auto b = Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(),
-              Kokkos::subview(elemRhs, ic, Kokkos::ALL()));
-          auto basisCoeffs_ = Kokkos::subview(basisCoeffs, ic, Kokkos::ALL());
-          auto serialBasisCoeffs = Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(),
-              basisCoeffs_);
+          Kokkos::deep_copy(A_device, Kokkos::subview(elemMat, ic, Kokkos::ALL(), Kokkos::ALL()));
+          Kokkos::deep_copy(A_host, A_device);
+          auto A = A_host;
+          Kokkos::deep_copy(b_device, Kokkos::subview(elemRhs, ic, Kokkos::ALL()));
+          Kokkos::deep_copy(b_host, b_device);
+          auto b = b_host;
+          auto basisCoeffsSubview = Kokkos::subview(basisCoeffs, ic, Kokkos::ALL());
+          Kokkos::deep_copy(basisCoeffsSubviewCopy_device, basisCoeffsSubview);
+          Kokkos::deep_copy(basisCoeffsSubviewCopy_host, basisCoeffsSubviewCopy_device);
```

Comment:
Do we really need multiple deep copies in the loop of Cells ? Too much overhead and consider a different refactoring approach.

Reply:
@mperego This might impose some inherent issue in designing the projection tools. Would you check and address the UVM issue on the projection tools from the class design ?